File: packages/typescript-sdk/src/composition/cost-budget.ts
import { CostBudgetExceededError } from "../errors";

export type CostBudgetOptions = {
  readonly maxCostMicrousd: number;
  readonly onWarning?: (current: number, max: number) => void;
  readonly warningThreshold?: number;
};

export type CostTracker = {
  readonly add: (costMicrousd: number) => void;
  readonly current: () => number;
  readonly remaining: () => number;
  readonly isExceeded: () => boolean;
};

export const createCostTracker = (options: CostBudgetOptions): CostTracker => {
  let currentCost = 0;
  const threshold = options.warningThreshold ?? 0.8;
  let warningFired = false;

  return {
    add: (costMicrousd) => {
      currentCost += costMicrousd;

      if (
        !warningFired &&
        options.onWarning &&
        currentCost >= options.maxCostMicrousd * threshold
      ) {
        warningFired = true;
        options.onWarning(currentCost, options.maxCostMicrousd);
      }

      if (currentCost >= options.maxCostMicrousd) {
        throw new CostBudgetExceededError({
          message: `Cost budget exceeded: ${currentCost} >= ${options.maxCostMicrousd} microusd`,
          currentCostMicrousd: currentCost,
          maxCostMicrousd: options.maxCostMicrousd,
        });
      }
    },
    current: () => currentCost,
    remaining: () => Math.max(0, options.maxCostMicrousd - currentCost),
    isExceeded: () => currentCost >= options.maxCostMicrousd,
  };
};

export const withCostBudget = <T>(
  fn: (tracker: CostTracker) => Promise<T>,
  options: CostBudgetOptions
): Promise<T> => {
  const tracker = createCostTracker(options);
  return fn(tracker);
};

File: packages/go-sdk/composition/cost_budget.go
package composition

import (
	"fmt"

	strait "github.com/strait-dev/go-sdk"
)

// CostBudgetOptions configures cost budget tracking.
type CostBudgetOptions struct {
	MaxCostMicrousd  int64
	OnWarning        func(current, max int64)
	WarningThreshold float64 // default 0.8
}

// CostTracker tracks accumulated costs against a budget.
type CostTracker struct {
	currentCost  int64
	options      CostBudgetOptions
	warningFired bool
}

// NewCostTracker creates a new cost tracker.
func NewCostTracker(opts CostBudgetOptions) *CostTracker {
	if opts.WarningThreshold == 0 {
		opts.WarningThreshold = 0.8
	}
	return &CostTracker{options: opts}
}

// Add adds cost and checks budget. Returns error if exceeded.
func (t *CostTracker) Add(costMicrousd int64) error {
	t.currentCost += costMicrousd

	threshold := int64(float64(t.options.MaxCostMicrousd) * t.options.WarningThreshold)
	if !t.warningFired && t.options.OnWarning != nil && t.currentCost >= threshold {
		t.warningFired = true
		t.options.OnWarning(t.currentCost, t.options.MaxCostMicrousd)
	}

	if t.currentCost >= t.options.MaxCostMicrousd {
		return &strait.CostBudgetExceededError{
			Message:             fmt.Sprintf("Cost budget exceeded: %d >= %d microusd", t.currentCost, t.options.MaxCostMicrousd),
			CurrentCostMicrousd: t.currentCost,
			MaxCostMicrousd:     t.options.MaxCostMicrousd,
		}
	}
	return nil
}

// Current returns the accumulated cost.
func (t *CostTracker) Current() int64 { return t.currentCost }

// Remaining returns the remaining budget.
func (t *CostTracker) Remaining() int64 {
	r := t.options.MaxCostMicrousd - t.currentCost
	if r < 0 {
		return 0
	}
	return r
}

// IsExceeded returns true if the budget is exceeded.
func (t *CostTracker) IsExceeded() bool { return t.currentCost >= t.options.MaxCostMicrousd }

// WithCostBudget creates a tracker and passes it to fn.
func WithCostBudget[T any](fn func(tracker *CostTracker) (T, error), opts CostBudgetOptions) (T, error) {
	tracker := NewCostTracker(opts)
	return fn(tracker)
}
